File: src/main/java/core/procedures/system/Name.java
package core.procedures.system;

import core.exceptions.WrongTypeException;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.INamed;

public final class Name extends AFn {

  public Name() {
    super(new FnArgsBuilder().minArgs(1).maxArgs(1));
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "name";
  }

  @Override
  public CharSequence apply1(Object arg) {
    if (arg instanceof INamed) {
      return ((INamed)arg).getName();
    }
    if (arg instanceof CharSequence) {
      return (CharSequence) arg;
    }
    throw new WrongTypeException(getName(), "String or Symbol or Keyword", arg);
  }
}

File: src/main/java/core/procedures/system/Name.kt
package core.procedures.system

import core.exceptions.WrongTypeException
import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.INamed

class Name : AFn<Any?, CharSequence>(name = "name", isPure = true, arity = Exactly(1)) {

    override operator fun invoke(arg: Any?) = when (arg) {
        is INamed -> arg.name
        is CharSequence -> arg
        else -> throw WrongTypeException(name, "String or Symbol or Keyword", arg)
    }
}
